File: implementations/fcp-ref/fcp_base/session/loop.py
"""
Main session loop — FCP §6.

Drives the cognitive cycle:
  drain io/inbox/ → consolidate → invoke CPE with growing chat_history
  → process tool_use → return tool_results → repeat

Context is assembled once at session start (system prompt + initial history from
session tail). Each cycle appends only the new stimulus to the in-memory
chat_history — the CPE never re-receives the boot manifest.

Session ends on session_close signal (CPE, SIL, or Operator).
"""

from __future__ import annotations

import json
import time
from typing import Any

from ..acp import make as acp_encode
from ..cpe.base import AdapterRef, CPEAdapter, CPEResponse, CPEError, CPEAuthError, CPERateLimitError
from ..cpe.models import get_context_window as _get_context_window
from ..dispatch import dispatch_tool_use
from ..mil import process_closure, summarize_session
from ..operator import (
    is_verbose as _is_verbose,
    is_compact_pending as _is_compact_pending,
    set_compact_pending as _set_compact_pending,
    is_endure_approved as _is_endure_approved,
    set_endure_approved as _set_endure_approved,
)
from ..session_mode import SessionMode, set_session_mode, get_session_mode, is_auto_session, is_main_session
from ..tools import build_tool_declarations as _tool_declarations
from ..store import Layout, append_jsonl, atomic_write, load_baseline, read_json
from .. import cmi_fmt as _cmi_fmt
from .. import ui
from .. import vital as _vital
from .context import build_boot_context, _rebuild_compact_history
from .history import (
    _drain_and_consolidate,
    _append_msg,
    _return_tool_result,
)
from .vlog import _vlog, _vlog_request, _vlog_cycle_summary

# Pure display helpers live in ui — import aliases for local use
_DIM   = ui.DIM
_RESET = ui.RESET
_GRAY  = ui.GRAY
_print_cpe_block       = ui.print_cpe_block
_vprint                = ui.vprint
_format_bytes          = ui.format_bytes
_compact_json          = ui.compact_json
_readline_with_history = ui.readline_with_history

_format_cmi_stimulus = _cmi_fmt.format_cmi_stimulus
_parse_env_data      = _cmi_fmt.parse_env_data
_cmi_indicator       = _cmi_fmt.cmi_indicator
_cmi_send_indicator  = _cmi_fmt.cmi_send_indicator
_envelope_to_text    = _cmi_fmt.envelope_to_text


# ---------------------------------------------------------------------------
# Loop Detection — Deterministic Fingerprinting
# ---------------------------------------------------------------------------

def _make_cycle_fingerprint(
    tool_calls: list[Any],
    tool_results: list[str],
) -> frozenset[tuple[str, str, str]]:
    """Create deterministic fingerprint of cycle (tool calls + results).

    Returns frozenset of (tool_name, input_hash, result_hash) tuples.
    Frozenset ensures order-independence and proper set comparison.

    Raises ValueError if tool_calls and tool_results counts don't match.
    """
    import hashlib

    if len(tool_calls) != len(tool_results):
        raise ValueError(
            f"Tool call/result count mismatch: {len(tool_calls)} calls, "
            f"{len(tool_results)} results"
        )

    fingerprints = []
    for call, result in zip(tool_calls, tool_results):
        # Hash input to avoid JSON stringification issues
        input_str = json.dumps(call.input, sort_keys=True, ensure_ascii=False)
        input_hash = hashlib.sha256(input_str.encode()).hexdigest()[:8]

        # Hash result similarly
        result_hash = hashlib.sha256(result.encode()).hexdigest()[:8]

        fingerprints.append((call.tool, input_hash, result_hash))

    return frozenset(fingerprints)


# ---------------------------------------------------------------------------
# Command Parsing Helper
# ---------------------------------------------------------------------------

def _parse_command(line: str) -> tuple[str, list[str]]:
    """Parse a command line into command name and arguments.

    Returns (command, args) tuple. Command is lowercased.
    Returns ('', []) if line is empty or doesn't start with '/'.

    Example:
    >>> _parse_command('/verbose --debug')
    ('/verbose', ['--debug'])
    >>> _parse_command('/')
    ('', [])
    >>> _parse_command('hello')
    ('', [])
    """
    stripped = line.strip()
    if not stripped.startswith('/'):
        return ('', [])

    parts = stripped.split()
    if not parts:
        return ('', [])

    command = parts[0].lower()
    args = parts[1:] if len(parts) > 1 else []
    return (command, args)


# ---------------------------------------------------------------------------
# CPE backoff policy
# ---------------------------------------------------------------------------

class _CPEBackoff:
    """Exponential backoff state for CPE transient errors."""

    MAX_CONSECUTIVE = 5
    INITIAL_SECS    = 1.0

    def __init__(self) -> None:
        self.consecutive = 0
        self.last_error_time = 0.0

    def reset(self) -> None:
        self.consecutive = 0
        self.last_error_time = 0.0

    def record_error(self) -> None:
        import time as _t
        self.consecutive += 1
        self.last_error_time = _t.time()

    @property
    def sleep_secs(self) -> float:
        return self.INITIAL_SECS * (2 ** (self.consecutive - 1))

    @property
    def exceeded(self) -> bool:
        return self.consecutive >= self.MAX_CONSECUTIVE


# ---------------------------------------------------------------------------
# Stimulus Collection and Input Handling
# ---------------------------------------------------------------------------

def _process_stimulus_and_input(
    layout: Layout,
    chat_history: list[dict[str, Any]],
    adapter_ref: AdapterRef,
) -> tuple[bool, bool, str]:
    """Collect stimuli from inbox and user input, return (stimulus_ready, session_closed, close_reason).

    Processes:
    1. Inbox drainage and consolidation
    2. User input handling (operator commands, text input)
    3. Command dispatch (/verbose, /exit, /new, /compact)

    Returns:
    - stimulus_ready: True if chat_history was updated with new content
    - session_closed: True if /exit or error occurred
    - close_reason: Reason for session close (if session_closed)
    """
    close_reason = ""
    stimulus_ready = False

    # Drain io/inbox/ → consolidate to session.jsonl
    inbox_envs = _drain_and_consolidate(layout)
    for env in inbox_envs:
        indicator = _cmi_indicator(env)
        if indicator:
            print(f"{_DIM}{indicator}{_RESET}")
        text = _envelope_to_text(env)
        if text:
            chat_history.append({"role": "user", "content": text})
            stimulus_ready = True

    # If no stimulus from inbox, wait for operator input before invoking CPE
    if not stimulus_ready:
        try:
            user_input = _readline_with_history("> ")
        except KeyboardInterrupt:
            print()
            return True, True, "operator_interrupt"
        except EOFError:
            return True, True, "operator_eof"

        stripped = user_input.strip()
        if not stripped:
            return False, False, ""

        # Platform commands — handle without invoking CPE
        if stripped.startswith("/"):
            from ..operator import handle_platform_command, _cmd_output
            handled = handle_platform_command(layout, stripped, adapter_ref=adapter_ref)
            if handled:
                cmd, _args = _parse_command(stripped)
                if _is_endure_approved():
                    _set_endure_approved(False)
                    return True, True, "endure_approved"
                if cmd in ("/exit", "/bye", "/close"):
                    return True, True, "operator_exit"
                if cmd in ("/new", "/clear", "/reset"):
                    return True, True, "operator_reset"
                if _is_compact_pending():
                    _set_compact_pending(False)
                    compact_msg = (
                        "[COMPACT_REQUEST] The operator has requested session compaction. "
                        "Generate a closure_payload now via fcp_mil to preserve your working context. "
                        "The session will continue after compaction — use session_handoff.next_steps "
                        "to describe where to resume."
                    )
                    _append_msg(layout, "fcp", compact_msg)
                    chat_history.append({"role": "user", "content": compact_msg})
                    return True, False, "compact_requested"
                return False, False, ""
            # unknown command
            cmd, _args = _parse_command(stripped)
            with _cmd_output():
                if cmd:
                    print(f"unknown command: {cmd}")
                else:
                    print("invalid command (must start with /)")
            return False, False, ""

        _append_msg(layout, "operator", user_input)
        chat_history.append({"role": "user", "content": stripped})
        stimulus_ready = True

    return stimulus_ready, False, ""


# ---------------------------------------------------------------------------
# Main session loop
# ---------------------------------------------------------------------------

def run_session(
    layout: Layout,
    adapter: CPEAdapter | AdapterRef,
    index: dict[str, Any],
    *,
    inject: list[dict[str, Any]] | None = None,
    greeting: bool = False,
    tools: list[dict[str, Any]] | None = None,
) -> str:
    """Run the cognitive session loop until session close (FCP §6).

    Executes the main session cycle:
    1. Build boot context (system prompt + history)
    2. Consume and inject first stimulus (FAP onboarding, post-evolution notices)
    3. Optionally inject additional stimuli
    4. Enter agentic loop: invoke CPE → dispatch tool calls → append results
    5. Continue until CPE requests close or operator exits
    6. Persist session store and compact if needed

    Args:
        layout: Entity store layout for session persistence.
        adapter: CPE adapter (Ollama, Claude, etc.) or reference for lazy initialization.
        index: Skill index for tool discovery and execution.
        inject: Optional list of ACP envelopes to prepend as initial stimuli (after first_stimuli).
        greeting: If True, inject SESSION_START stimulus to wake CPE for greeting.
        tools: Optional pre-built tool declarations. If None, auto-discovered from index.

    Returns:
        str: Close reason code (e.g., "session_close", "operator_exit", "max_turns").

    Raises:
        CPEError: If CPE invocation fails (model unavailable, API error).
        ExecError: If tool execution fails or security check blocks an operation.
        BootError: If critical conditions (session token, session store) are unmet.
    """
    if tools is None:
        tools = _tool_declarations(layout, index)
    adapter_ref = adapter if isinstance(adapter, AdapterRef) else AdapterRef(adapter)
    _set_endure_approved(False)


    # --- Build system prompt and initial chat history once at session start ---
    system, chat_history = build_boot_context(layout, index)
    _vlog("fcp", f"boot context: system={len(system)} chars, history={len(chat_history)} msgs")

    from ..stimuli import pop_stimulus
    first_stimuli_injected = False
    # Consume first_stimuli if present (e.g. FAP onboarding, post-evolution notice)
    fs = pop_stimulus(layout)
    if fs:
        msg = str(fs.get("message", ""))
        if msg:
            env = acp_encode(env_type="MSG", source="fcp",
                             data={"type": "FIRST_STIMULI", "source": fs.get("source", "fcp"), "msg": msg})
            append_jsonl(layout.session_store, env)
            chat_history.append({"role": "user", "content": msg})
            first_stimuli_injected = True
            _vlog("fcp", f"first_stimuli injected (source={fs.get('source')})")

    if inject:
        for env in inject:
            append_jsonl(layout.session_store, env)
            text = _envelope_to_text(env)
            if text:
                chat_history.append({"role": "user", "content": text})

    close_reason = "session_close"
    cycle = 0
    compact_in_progress = False
    stimulus_ready = bool(greeting or inject or first_stimuli_injected)
    tokens_used = 0

    # loop detection: track last N cycle fingerprints (each a frozenset of (tool, input_json, result) tuples)
    _loop_window: list[Any] = []
    _LOOP_THRESHOLD = 3

    # CPE error handling — exponential backoff for transient failures
    _backoff = _CPEBackoff()

    # Vital Check state — triggers on cycle_threshold or interval_seconds
    _baseline = None
    _vital_state = None
    _ctx_window = 0
    _cpe_backend = ""
    try:
        from ..formats import StructuralBaseline
        _baseline = StructuralBaseline.from_dict(read_json(layout.baseline))
        _cpe_backend = _baseline.cpe.backend
        _session_id = ""
        if layout.session_token.exists():
            _session_id = str(read_json(layout.session_token).get("session_id", ""))
        _vital_state = _vital.VitalCheckState(session_id=_session_id)
    except FileNotFoundError:
        _vlog("fcp", "baseline file not found — vital check disabled")
    except json.JSONDecodeError as e:
        _vlog("fcp", f"baseline file corrupted ({e}) — vital check disabled")
    except Exception as e:
        _vlog("fcp", f"baseline load error ({e}) — vital check disabled")

    while True:
        # If pre-loaded stimulus (first_stimuli, greeting, inject), skip input collection
        if not stimulus_ready:
            stimulus_ready, should_close, close_reason = _process_stimulus_and_input(
                layout, chat_history, adapter_ref
            )
            if should_close:
                if close_reason == "compact_requested":
                    # Compact was requested; set flag and inject message
                    compact_in_progress = True
                    stimulus_ready = True
                else:
                    # Session should close
                    break

        if not stimulus_ready:
            continue

        stimulus_ready = False
        cycle += 1
        cycle_start_time = time.time()
        _vlog_request(system, chat_history, tools, cycle)

        # invoke CPE (adapter_ref.current may be swapped mid-session via /model)
        try:
            response = adapter_ref.current.invoke(system, chat_history, tools)
            _backoff.reset()
        except CPEAuthError as exc:
            # Authentication errors are not retryable — exit immediately
            err_msg = f"CPE authentication failed: {str(exc)}"
            print(f"\n{_DIM}  [fcp] {err_msg}{_RESET}")
            _append_msg(layout, "fcp", err_msg)
            _vlog("fcp", f"auth error: {err_msg}")
            close_reason = "cpe_auth_error"
            break
        except CPERateLimitError as exc:
            _backoff.record_error()
            err_msg = f"CPE rate limited (attempt {_backoff.consecutive}): {str(exc)}"
            print(f"\n{_DIM}  [fcp] {err_msg} (backoff {_backoff.sleep_secs:.1f}s){_RESET}")
            _append_msg(layout, "fcp", f"{err_msg} — retrying in {_backoff.sleep_secs:.0f}s")
            _vlog("fcp", f"rate limit backoff: {_backoff.sleep_secs}s")
            if _backoff.exceeded:
                close_reason = "cpe_rate_limit_exceeded"
                break
            time.sleep(_backoff.sleep_secs)
            stimulus_ready = False
            continue
        except CPEError as exc:
            _backoff.record_error()
            err_msg = f"CPE error (attempt {_backoff.consecutive}): {str(exc)}"
            print(f"\n{_DIM}  [fcp] {err_msg} (backoff {_backoff.sleep_secs:.1f}s){_RESET}")
            _append_msg(layout, "fcp", f"{err_msg} — retrying in {_backoff.sleep_secs:.0f}s")
            _vlog("fcp", f"cpe error backoff: {_backoff.sleep_secs}s")
            if _backoff.exceeded:
                close_reason = "cpe_error_max_retries"
                break
            time.sleep(_backoff.sleep_secs)
            stimulus_ready = False
            continue
        except Exception as exc:
            _backoff.record_error()
            err_msg = f"CPE unexpected error (attempt {_backoff.consecutive}): {str(exc)}"
            print(f"\n{_DIM}  [fcp] {err_msg}{_RESET}")
            _append_msg(layout, "fcp", err_msg)
            _vlog("fcp", f"unexpected error: {err_msg}")
            if _backoff.exceeded:
                close_reason = "cpe_error_max_retries"
                break
            stimulus_ready = False
            continue
        cycle_elapsed = time.time() - cycle_start_time
        tokens_used = tokens_used + response.input_tokens + response.output_tokens

        # add CPE response to chat history and display status
        if response.tool_use_calls and not _is_verbose():
            n = len(response.tool_use_calls)
            print(f"\n{_DIM}  ○ cycle {cycle} — {n}x tool call{'s' if n != 1 else ''}{_RESET}")
        _model_label = getattr(adapter_ref.current, "_model", "")
        _ctx_window = _get_context_window(_cpe_backend, _model_label)
        if response.text:
            _append_msg(layout, "cpe", response.text)
            _print_cpe_block(response.text, _model_label, response.input_tokens, response.output_tokens, _ctx_window)
            chat_history.append({"role": "assistant", "content": response.text})
        if response.tool_use_calls:
            # Always append an empty assistant sentinel when there are tool calls.
            # Adapters detect tool-result turns by checking that the preceding assistant
            # turn has empty content. Without this sentinel, text+tool_calls responses
            # leave a non-empty assistant turn, causing tool results to be treated as
            # plain text in subsequent cycles ("soluço" / hiccup bug).
            chat_history.append({"role": "assistant", "content": ""})

        # process tool_use calls — fcp_mil before fcp_exec before fcp_sil (per spec)
        tool_calls = sorted(
            response.tool_use_calls,
            key=lambda c: 0 if c.tool == "fcp_mil" else (1 if c.tool == "fcp_exec" else 2),
        )

        session_closed = False
        tool_results: list[str] = []
        tool_log_lines: list[dict[str, Any]] = []

        for i, call in enumerate(tool_calls):
            tool_start = time.time()
            result, closed = dispatch_tool_use(layout, call, index)
            tool_elapsed = time.time() - tool_start

            # Accumulate tool execution info (printed later in cycle summary)
            is_last = i == len(tool_calls) - 1
            input_size = _format_bytes(len(json.dumps(call.input, ensure_ascii=False)))
            result_size = _format_bytes(len(json.dumps(result, ensure_ascii=False)))
            status = "OK" if isinstance(result, dict) and result.get("error") is None else "FAIL"
            timing_ms = tool_elapsed * 1000

            tool_log_lines.append({
                "tool": call.tool,
                "is_last": is_last,
                "input": call.input,
                "output": result,
                "input_size": input_size,
                "result_size": result_size,
                "status": status,
                "timing_ms": timing_ms,
            })

            _return_tool_result(layout, call.id, call.tool, result)
            # Serialize result once and reuse for chat history
            # This avoids triple JSON serialization (verbose logging, chat history, loop detection)
            result_str = json.dumps(result, ensure_ascii=False)
            tool_results.append(f"[{call.tool}] {result_str}")
            if call.tool == "cmi_send":
                _cmi_send_indicator(call.input, result)
            if closed:
                close_reason = "session_close"
                session_closed = True
            if _is_endure_approved():
                close_reason = "endure_approved"
                session_closed = True

        # Print cycle summary: [DISPATCH] + [← CPE] in correct order
        _budget_pct = _baseline.context_window_budget_pct if _baseline else 0
        _vlog_cycle_summary(response, cycle_elapsed, tool_log_lines, _ctx_window, _budget_pct, _cpe_backend, _model_label)

        if session_closed:
            break

        # tool results go into chat history as full payloads.
        # result_recall remains available as fallback for results from previous sessions.
        if tool_results:
            chat_history.append({"role": "user", "content": "\n".join(tool_results)})
            stimulus_ready = True  # tool results need a follow-up CPE cycle

        # --- loop detection: same set of (tool, input, result) tuples repeated >= threshold ---
        if tool_calls:
            try:
                # Create deterministic fingerprint (handles JSON stringification issues)
                cycle_fingerprint = _make_cycle_fingerprint(tool_calls, tool_results)
                _loop_window.append(cycle_fingerprint)
                if len(_loop_window) > _LOOP_THRESHOLD:
                    _loop_window.pop(0)

                # Detect loop: same fingerprint repeated THRESHOLD times
                if len(_loop_window) == _LOOP_THRESHOLD and len(set(_loop_window)) == 1:
                    _loop_window.clear()
                    tools_repr = ", ".join(c.tool for c in tool_calls)
                    intervention = (
                        f"[FCP] Loop detected: the same tool call(s) ({tools_repr}) returned "
                        f"identical results {_LOOP_THRESHOLD} times in a row. "
                        "Stop and report the situation to the Operator. Do not retry."
                    )
                    _vlog("fcp", f"loop detected: {tools_repr}")
                    _append_msg(layout, "fcp", intervention)
                    chat_history.append({"role": "user", "content": intervention})
                    stimulus_ready = True
            except ValueError as e:
                # Tool call/result count mismatch (shouldn't happen, but log if it does)
                _vlog("fcp", f"loop detection skipped: {e}")
                _loop_window.clear()
        else:
            _loop_window.clear()

        # Vital Check — tick counter; run if either trigger threshold is reached
        if _vital_state is not None and _baseline is not None:
            _vital.tick(_vital_state)
            if _vital.should_run(_vital_state, _baseline):
                _vital.run(layout, _baseline, _vital_state, tokens_used, _cpe_backend, _model_label)

        # compact: if closure_payload was written during this cycle, execute Stage 1
        # and rebuild chat_history with the condensed context
        if compact_in_progress and layout.pending_closure.exists():
            compact_in_progress = False
            _vlog("fcp", "compact: processing closure payload")
            process_closure(layout)
            chat_history[:] = _rebuild_compact_history(layout, index, system)
            summarize_session(layout)
            _vlog("fcp", f"compact: done — history={len(chat_history)} msgs")
            print(f"\n{_DIM}  [fcp] session compacted{_RESET}")

    _vlog("fcp", f"session closed — reason: {close_reason}")
    return close_reason

File: implementations/fcp-ref-ts/src/session/loop.ts
// Session loop — main cognitive cycle.
// Iterates: drain inbox → operator prompt → CPE invoke → tool dispatch → heartbeat → repeat.
// EXCEPTION: ~220 lines — tool dispatch and message persistence are tightly coupled
// to the main loop invariant and cannot be split without losing readability.
import * as path from 'node:path'
import * as nodefs from 'node:fs/promises'
import { appendJsonl, readJson, fileExists, ensureDir, writeJson } from '../store/io.js'
import { drainInbox } from './inbox.js'
import { buildContext } from './context.js'
import { estimateTokens, checkBudget } from './budget.js'
import { makeFingerprint } from './fingerprint.js'
import { SESSION_CLOSE_SIGNAL, COMPACT_SESSION_SIGNAL } from '../sil/sil.js'
import { ClosurePayloadSchema } from '../types/formats/memory.js'
import type { SessionOptions, LoopResult, CycleState } from '../types/session.js'
import type { CPEMessage, ToolResultBlock } from '../types/cpe.js'
import type { HeartbeatResult } from '../types/sil.js'

const MAX_TOOL_CYCLES = 50

async function writeOperatorNotification(
  layout: import('../types/store.js').Layout,
  level: 'degraded' | 'critical',
  result: HeartbeatResult,
): Promise<void> {
  await ensureDir(layout.state.operatorNotifications)
  const ts   = new Date().toISOString().replace(/[:.]/g, '-')
  const file = path.join(layout.state.operatorNotifications, `heartbeat-${level}-${ts}.json`)
  const failing = result.vitals.filter(v => !v.ok)
  await writeJson(file, { level, ts: result.ts, cycleCount: result.cycleCount, budgetPct: result.budgetPct, failing })
}

export async function runSessionLoop(opts: SessionOptions): Promise<LoopResult> {
  const { layout, baseline, cpe, policy, tools, logger, io, sessionId, contextMessages, heartbeat } = opts
  const contextWindow = opts.contextWindow > 0 ? opts.contextWindow : baseline.contextWindow.fallbackTokens
  const log = logger.child({ module: 'session', sessionId })

  const toolMap = new Map(tools.map(t => [t.name, t]))
  const firstWriteDone: { value: boolean } = { value: false }
  const execCtx = { layout, baseline, logger, sessionId, sessionMode: 'main' as const, policy, io, firstWriteDone }
  const { system } = await buildContext(layout)

  // MIL/SIL tools bypass the operator gate — they are fundamental to system operation.
  const MIL_SIL_TOOLS = new Set([
    'fcp_memory_recall', 'fcp_memory_write', 'fcp_closure_payload',
    'fcp_evolution_proposal', 'fcp_session_close',
  ])

  const messages: CPEMessage[] = [...(contextMessages ?? [])]
  const fingerprints: string[] = []
  let cycle: CycleState = { cycleNum: 0, inputTokens: 0, fingerprint: '' }
  // Set to true when the SIL compact check fires — prevents re-injection on next cycle.
  let compactRequested = false

  log.info('session:start')

  try {
    while (true) {
      io.emit({ type: 'cycle_start', cycleNum: cycle.cycleNum + 1 })

      // ── Drain async inbox ────────────────────────────────────────────────
      let compactSignalReceived = false
      for (const msg of await drainInbox(layout)) {
        const text = typeof msg.content === 'string' ? msg.content : JSON.stringify(msg.content)
        if (text === COMPACT_SESSION_SIGNAL) {
          compactSignalReceived = true
          continue
        }
        messages.push(msg)
        await appendJsonl(layout.memory.sessionJsonl, { type: 'message', role: 'user', content: text, ts: new Date().toISOString() })
        io.emit({ type: 'operator_msg', content: text })
      }

      // ── Compact signal: inject CPE instruction ───────────────────────────
      if (compactSignalReceived) {
        const compactInstruction = 'SYSTEM: Context window is critically full. Stop current work and immediately call fcp_closure_payload to save state. The session will be restarted for compaction.'
        messages.push({ role: 'user', content: compactInstruction })
        await appendJsonl(layout.memory.sessionJsonl, { type: 'message', role: 'user', content: compactInstruction, ts: new Date().toISOString() })
        log.warn('session:compact_requested')
      }

      // ── Operator input (when idle) ───────────────────────────────────────
      const lastRole = messages.at(-1)?.role
      if (messages.length === 0 || lastRole === 'assistant') {
        const text = await io.prompt()
        const trimmed = text.trim()
        if (!trimmed) continue

        messages.push({ role: 'user', content: trimmed })
        await appendJsonl(layout.memory.sessionJsonl, { type: 'message', role: 'user', content: trimmed, ts: new Date().toISOString() })
        io.emit({ type: 'operator_msg', content: trimmed })
      }

      // ── Budget check (operator-visible, based on 95% of window) ─────────
      const tokens = cycle.inputTokens > 0 ? cycle.inputTokens : estimateTokens(messages)
      const operatorMax = Math.round(contextWindow * 0.95)
      const budget = checkBudget(tokens, operatorMax, baseline.contextWindow.criticalPct, baseline.contextWindow.warnPct)
      // Emit token update so TUI can display in/out counts and budget %
      io.emit({
        type:         'token_update',
        inputTokens:  tokens,
        outputTokens: 0,   // output tokens only available after CPE response
        budgetPct:    budget.usedPct,
      })
      if (budget.status === 'critical') {
        io.emit({ type: 'session_close', reason: 'budget_critical' })
        return { closed: 'forced', reason: 'budget_critical' }
      }

      // ── CPE invoke ───────────────────────────────────────────────────────
      io.emit({ type: 'cpe_invoke' })
      const resp = await cpe.invoke({
        system,
        messages,
        tools: tools.map(t => ({
          name:         t.name,
          description:  t.description,
          input_schema: t.inputSchema,
        })),
      })
      cycle = {
        cycleNum:    cycle.cycleNum + 1,
        inputTokens: resp.usage.inputTokens,
        fingerprint: makeFingerprint(resp.toolUses),
      }
      io.emit({ type: 'cpe_response', content: resp.content, toolUses: resp.toolUses })
      io.emit({
        type:         'token_update',
        inputTokens:  resp.usage.inputTokens,
        outputTokens: resp.usage.outputTokens,
        budgetPct:    Math.round((resp.usage.inputTokens / operatorMax) * 100),
      })

      // ── End turn — push assistant message, wait for operator ─────────────
      if (resp.stopReason === 'end_turn' || resp.toolUses.length === 0) {
        messages.push({ role: 'assistant', content: resp.content })
        await appendJsonl(layout.memory.sessionJsonl, { type: 'message', role: 'assistant', content: resp.content, ts: new Date().toISOString() })
        fingerprints.length = 0
        continue
      }

      // ── Tool cycle ───────────────────────────────────────────────────────
      if (cycle.cycleNum >= MAX_TOOL_CYCLES) {
        log.warn('session:max-cycles', { cycleNum: cycle.cycleNum })
        return { closed: 'forced', reason: 'critical_condition' }
      }
      if (fingerprints.includes(cycle.fingerprint)) {
        log.warn('session:loop-detected', { fingerprint: cycle.fingerprint })
        return { closed: 'forced', reason: 'critical_condition' }
      }
      fingerprints.push(cycle.fingerprint)

      // Push assistant tool-use block
      messages.push({ role: 'assistant', content: resp.toolUses })
      await appendJsonl(layout.memory.sessionJsonl, { type: 'message', role: 'assistant', content: resp.toolUses, ts: new Date().toISOString() })

      // Dispatch each tool use
      const results: ToolResultBlock[] = []
      let sessionCloseTriggered = false
      for (const tu of resp.toolUses) {
        io.emit({ type: 'tool_dispatch', skillName: tu.name, input: tu.input })
        const handler = toolMap.get(tu.name)
        if (!handler) {
          results.push({ type: 'tool_result', tool_use_id: tu.id, content: `Unknown tool: ${tu.name}` })
          io.emit({ type: 'tool_result', skillName: tu.name, result: { ok: false, error: 'unknown' } })
          continue
        }
        // MIL/SIL tools bypass the operator gate — gate is owned by EXEC tool handlers.
        const approved = MIL_SIL_TOOLS.has(tu.name) ? 'bypass' : 'exec'
        const result = await handler.execute(tu.input, execCtx)
        results.push({ type: 'tool_result', tool_use_id: tu.id, content: result.ok ? result.output : `Error: ${result.error}` })
        io.emit({ type: 'tool_result', skillName: tu.name, result })
        await appendJsonl(layout.memory.sessionJsonl, { type: 'tool_execution', tool: tu.name, approved, ts: new Date().toISOString() })
        if (result.ok && result.output === SESSION_CLOSE_SIGNAL) {
          sessionCloseTriggered = true
        }
      }

      // Push tool results as user message
      messages.push({ role: 'user', content: results })
      await appendJsonl(layout.memory.sessionJsonl, { type: 'message', role: 'user', content: results, ts: new Date().toISOString() })

      if (sessionCloseTriggered) break

      // ── Heartbeat ────────────────────────────────────────────────────────
      if (heartbeat && await heartbeat.shouldRun(cycle.cycleNum)) {
        const hbResult = await heartbeat.run(cycle.cycleNum, cycle.inputTokens, contextWindow)
        const criticals = hbResult.vitals.filter(v => !v.ok && v.severity === 'critical')
        const degraded  = hbResult.vitals.filter(v => !v.ok && v.severity === 'degraded')

        // compact_session check fires at 97% — inject inbox message for next cycle
        const compactVital = criticals.find(v => v.check === 'compact_session')
        if (compactVital && !compactRequested) {
          compactRequested = true
          await ensureDir(layout.io.inbox)
          const ts   = new Date().toISOString().replace(/[:.]/g, '-')
          const file = path.join(layout.io.inbox, `compact-${ts}.json`)
          await writeJson(file, COMPACT_SESSION_SIGNAL)
          const compactMsg = !compactVital.ok ? compactVital.message : ''
          log.warn('session:compact_signal_injected', { pct: compactMsg })
          // Remove compact from criticals list so the other checks still run
          const otherCriticals = criticals.filter(v => v.check !== 'compact_session')
          if (otherCriticals.length > 0) {
            await writeOperatorNotification(layout, 'critical', hbResult)
            io.emit({ type: 'session_close', reason: 'critical_condition' })
            log.warn('session:heartbeat:critical', { vitals: otherCriticals.map(v => v.check) })
            return { closed: 'forced', reason: 'critical_condition' }
          }
        } else if (criticals.length > 0) {
          await writeOperatorNotification(layout, 'critical', hbResult)
          await nodefs.unlink(layout.state.sentinels.sessionToken).catch(() => undefined)
          io.emit({ type: 'session_close', reason: 'critical_condition' })
          log.warn('session:heartbeat:critical', { vitals: criticals.map(v => v.check) })
          return { closed: 'forced', reason: 'critical_condition' }
        }
        if (degraded.length > 0) {
          await writeOperatorNotification(layout, 'degraded', hbResult)
          log.warn('session:heartbeat:degraded', { vitals: degraded.map(v => v.check) })
        }
      }
    }
  } catch (e: unknown) {
    io.emit({ type: 'error', error: e })
    log.error('session:error', { err: String(e) })
    return { closed: 'error', error: e }
  }

  // ── Normal / compact close ───────────────────────────────────────────────
  io.emit({ type: 'session_close', reason: 'normal' })
  log.info('session:close', { cycles: cycle.cycleNum, compact: compactRequested })

  const raw = await fileExists(layout.state.pendingClosure)
    ? await readJson(layout.state.pendingClosure)
    : null
  const parsed = ClosurePayloadSchema.safeParse(raw)
  if (!parsed.success) {
    log.warn('session:no-closure-payload')
    return { closed: 'forced', reason: 'critical_condition' }
  }
  return { closed: 'normal', closurePayload: parsed.data, compact: compactRequested }
}
